L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+    // each trip corresponds to a f/w clock tick
+    // done if no states to process left, taking as much time as needed
+    while (!stream.empty() || !stack.empty() || !all_of(delay.begin(), delay.end(), [](const State* state){ return state == nullptr; })) {
+      State* state = pop_front(stream);
+      if (!state)
+        state = pop_front(stack);
+      streamOutput.push_back(state);
+      if (!state || !state->stub() || state->layer() != layer_)
+        state = nullptr;
+      if (state != nullptr)
+        comb(state);
+      delay.push_back(state);
+      state = pop_front(delay);
+      if (state != nullptr)
+        stack.push_back(state);
+    }
+    stream = streamOutput;
+    for (State*& state : stream) {
+      if (!state || !state->stub() || state->layer() != layer_ )
+        continue;
+      update(state);
+    }
+  }
+
+  // repicks combinatoric stubs for state
+  void KalmanFilter::comb(State*& state) {
+    // picks next stub on layer
+    StubKFin* stub = state->stub();
+    const int layer = stub->layer();
+    TrackKFin* track = state->track();
+    const vector<StubKFin*>& stubs = track->layerStubs(layer);
+    const TTBV& hitPattern = state->hitPattern();
+    const int pos = distance(stubs.begin(), find(stubs.begin(), stubs.end(), stub)) + 1;
+    StubKFin* stubNext = nullptr;
+    if (pos != (int)stubs.size())
+      stubNext = stubs[pos];
+    // picks next stub on different layer, nullifies state if skipping layer is not valid
+    else {
+      bool valid(true);
+      // having already maximum number of added layers
+      if (hitPattern.count() == setup_->kfMaxLayers())
+        valid = false;
+      // not enough layers remain after skipping
+      if (hitPattern.count() + track->hitPattern().count(layer + 1, setup_->numLayers()) < setup_->kfMaxLayers())
+        valid = false;
+      if (valid) {
+        // pick next stub on next populated layer
+        for (int nextLayer = layer_ + 1; nextLayer < setup_->numLayers(); nextLayer++) {
+          if (track->hitPattern(nextLayer)) {
+            stubNext = track->layerStub(nextLayer);
+            break;
+          }
+        }
+      }
+    }
+    if (stubNext) {
+      states_.emplace_back(state, stubNext);
+      state = &states_.back();
+    } else
+      state = nullptr;
+  }
+
+  // best state selection
+  void KalmanFilter::accumulator(deque<State*>& stream) {
+    // accumulator delivers contigious stream of best state per track
+    // remove gaps and not final states
+    stream.erase(remove_if(stream.begin(), stream.end(), [this](State* state){ return !state || state->hitPattern().count() < setup_->kfMaxLayers(); }), stream.end());
+    /*auto toLessPS = [this](State* state) {
+      int nPS(0);
+      for (const StubKF& stub : state->stubs())
+        if (setup_->psModule(stub.ttStubRef()))
+          nPS++;
+      return nPS < 2;
+    };
+    stream.erase(remove_if(stream.begin(), stream.end(), toLessPS), stream.end());*/
+    // update chi2
```

Comment:
Is this comment wrong? State::finish() calculates variables related to track quality, but not chi2.

Reply:
done